runtime/common/src/assigned_slots.rs

```diff
@@ -0,0 +1,1281 @@
+		ParaId,
+		Option<ParachainTemporarySlot<T::AccountId, LeasePeriodOf<T>>>,
+		ValueQuery,
+	>;
+
+	/// Number of assigned temporary slots.
+	#[pallet::storage]
+	#[pallet::getter(fn temporary_slot_count)]
+	pub type TemporarySlotCount<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	/// Number of active temporary slots in current slot lease period.
+	#[pallet::storage]
+	#[pallet::getter(fn active_temporary_slot_count)]
+	pub type ActiveTemporarySlotCount<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A para was assigned a permanent parachain slot
+		PermanentSlotAssigned(ParaId),
+		/// A para was assigned a temporary parachain slot
+		TemporarySlotAssigned(ParaId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The specified parachain or parathread is not registered.
+		ParaDoesntExist,
+		/// Not a parathread.
+		NotParathread,
+		/// Cannot upgrade parathread.
+		CannotUpgrade,
+		/// Cannot downgrade parachain.
+		CannotDowngrade,
+		/// Permanent or Temporary slot already assigned.
+		SlotAlreadyAssigned,
+		/// Permanent or Temporary slot has not been assigned.
+		SlotNotAssigned,
+		/// An ongoing lease already exists.
+		OngoingLeaseExists,
+		// Maximum number of permanent slots exceeded
+		MaxPermanentSlotsExceeded,
+		// Maximum number of temporary slots exceeded
+		MaxTemporarySlotsExceeded,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			if let Some((lease_period, first_block)) = Self::lease_period_index(n) {
+				// If we're beginning a new lease period then handle that.
+				if first_block {
+					return Self::manage_lease_period_start(lease_period)
+				}
+			}
+
+			// We didn't return early above, so we didn't do anything.
+			0
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Assign a permanent parachain slot and immediately create a lease for it.
+		#[pallet::weight((1_000 + <T as slots::Config>::WeightInfo::force_lease(), DispatchClass::Operational))]
+		pub fn assign_perm_parachain_slot(origin: OriginFor<T>, id: ParaId) -> DispatchResult {
+			T::AssignSlotOrigin::ensure_origin(origin)?;
+
+			let manager = T::Registrar::manager_of(id).ok_or(Error::<T>::ParaDoesntExist)?;
+
+			ensure!(T::Registrar::is_parathread(id), Error::<T>::NotParathread,);
+
+			ensure!(
+				!Self::has_permanent_slot(id) && !Self::has_temporary_slot(id),
+				Error::<T>::SlotAlreadyAssigned
+			);
+
+			let current_lease_period: T::BlockNumber = Self::current_lease_period_index();
+			ensure!(
+				!T::Leaser::already_leased(
+					id,
+					current_lease_period,
+					// Check current lease & next one
+					current_lease_period.saturating_add(
+						T::BlockNumber::from(2u32)
+							.saturating_mul(T::PermanentSlotLeasePeriodLength::get().into())
+					)
+				),
+				Error::<T>::OngoingLeaseExists
+			);
```

Comment:
You are already placing a hard dependency on the `slots` pallet, so probably you can do a more efficient check here directly with the storage struct.

Reply:
Using `contains-key` on the leases storage map would not be enough .. then I'd have to re-implement some of the logic that's nicely encapsulated (and tested) in `T::Leaser::already-leased` function. Also this call shouldn't be called that often, so I'd go for correctness over further optimisation, what you think ?